service/storage_service.cc

```diff
@@ -2575,9 +2584,105 @@ class topology_coordinator {
+
+            auto timeout = netw::messaging_service::clock_type::now() + wait_for_live_nodes_timeout;
+
+            abort_source as;
+            auto request_abort = [&as] () mutable noexcept {
+                as.request_abort();
+            };
+            auto t = timer<lowres_clock>(request_abort);
+            t.arm(timeout);
+            auto sub = _as.subscribe(request_abort);
+
+            auto node_stats = co_await ser::storage_service_rpc_verbs::send_table_load_stats(&_messaging,
+                netw::msg_addr(id2ip(dst)),
+                as,
+                raft::server_id(dst.uuid()));
+
+            dc_stats += node_stats;
+        });
+
+        for (auto& [table_id, table_stats] : dc_stats.tables) {
+            co_await coroutine::maybe_yield();
+
+            auto& t = _db.find_column_family(table_id);
+            auto& rs = t.get_effective_replication_map()->get_replication_strategy();
+            if (!rs.uses_tablets()) {
+                continue;
+            }
+            const auto* nts_ptr = dynamic_cast<const locator::network_topology_strategy*>(&rs);
+            if (!nts_ptr) {
+                on_internal_error(slogger, "Cannot convert replication_strategy that uses tablets into network_topology_strategy");
+            }
+
+            auto rf_for_this_dc = nts_ptr->get_replication_factor(dc);
+            if (rf_for_this_dc <= 0) {
+                continue;
+            }
+            // Takes into account the RF of each DC, so stats will still make sense when aggregated
+            // from multiple DCs. This allows tablet allocator to compute an average tablet size
+            // by dividing total size in bytes by tablet count.
+            table_stats.size_in_bytes /= rf_for_this_dc;
+            slogger.info("raft topology: Refreshed table load stats for DC {}, table={}, RF={}, size_in_bytes={}, split_ready_phase={}",
+                          dc, table_id, rf_for_this_dc, table_stats.size_in_bytes, table_stats.split_ready_phase);
+        }
+
+        stats += dc_stats;
+    }
+
+    co_return std::move(stats);
+}
+
+future<> topology_coordinator::start_tablet_load_stats_refresher() {
+    auto can_proceed = [this] { return !_async_gate.is_closed() && !_as.abort_requested(); };
+    while (can_proceed()) {
+        bool sleep = true;
+        try {
+            _tablet_load_stats = make_lw_shared<const locator::load_stats>(co_await refresh_tablet_load_stats());
+            _topo_sm.event.broadcast(); // wake up load balancer.
+        } catch (raft::request_aborted&) {
+            slogger.debug("raft topology: Tablet load stats refresher aborted");
+            sleep = false;
+        } catch (seastar::abort_requested_exception) {
+            slogger.debug("raft topology: Tablet load stats refresher aborted");
+            sleep = false;
+        } catch (...) {
+            slogger.warn("Found error while refreshing load stats for tablets: {}, retrying...", std::current_exception());
```

Comment:
Ok, you handle them here, but if a node is down, the whole thing stops working.

With a global tablets map, it just has stale data.

Reply:
I intended to fix it later. But maybe it's better to do it now. We cannot trust the split status if any node is down, but the we can initiate new splits if a small percentage of nodes is down. Otherwise, inference of avg size could be way off. What should be this percentage? 1%? 5%? (to a minimum of 1 node).

Reply:
Consider we have 3 nodes and RF 3

let's say there are 2 tablets.

each node has 1 tablet replica for each of the 2 tablets.

if we lose 1 node, we still have access to 4 tablet replicas

if RPC handler counts tablet replicas, we can still normalize table_total_size by RF

in a normal condition, calculation would be:
`    table_total_size / RF=3`
    
but with loss of 1 node, we missed 2 tablet replicas

```
expected_tablet_replicas = RF * tablet_count = 6
found_tablet_replicas = (accumulated by RPC handler) = 4
```

To find actual RF (available at the time), we can do:

`available_RF = found_tablet_replicas / expected_tablet_replicas * RF = 2`

so normalization of table_total_size can be done against available_RF

as for the failure rate accepted, we can probably start with something like:

`failure_rate_accepted = std::max(floor(nodes.size() * 0.75), 1)`

Reply:
You say it won't help since the split speed is determined by the slowest replica, right?

BTW, double checked that nodes that come back up will register tables with pending splits. That happens when loading tablet metadata on startup and replicating the token metadata into all shards.

Reply:
Once we get rid of power-of-two constraint, then everything changes.

Reply:
Better than double checking is to add a test for it :-) i will do.